### PR TITLE
[Checkpoint] wait for checkpoint service to stop during reconfig

### PR DESCRIPTION
**Pull request by mwtian**
## Description 

Currently during reconfig, `CheckpointService` tasks, including `CheckpointBuilder` and `CheckpointAggregator`, are [notified to shut down](https://github.com/MystenLabs/sui/blob/b1540cdb2019f0501d2cd7f6dad208f65a66b6d2/crates/sui-node/src/lib.rs#L1551). But reconfig does not wait for them to finish shutting down. There can be a race between the reconfig loop proceeding to [drop the epoch db handle](https://github.com/MystenLabs/sui/blob/b1540cdb2019f0501d2cd7f6dad208f65a66b6d2/crates/sui-node/src/lib.rs#L1633), while `CheckpointBuilder` tries to [read from epoch db when creating a new checkpoint](https://github.com/MystenLabs/sui/blob/b1540cdb2019f0501d2cd7f6dad208f65a66b6d2/crates/sui-core/src/checkpoints/mod.rs#L1378). The race can result in panics.

## Test plan 

CI
Simulation

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

**Comment by halfprice** [comment]
Discussed this with @mwtian in the office the other day. Can we make sure it is actually safe to cancel `make_checkpoint` and aggregator's `run_and_notify`, and won't leave inconsistent state in the memory? Maybe consider using fail_points to cancel these tasks while it is running, and see if it causes any errors.

**Comment by mystenmark** [comment]
> Discussed this with @mwtian in the office the other day. Can we make sure it is actually safe to cancel `make_checkpoint` and aggregator's `run_and_notify`, and won't leave inconsistent state in the memory? Maybe consider using fail_points to cancel these tasks while it is running, and see if it causes any errors.

I had the same question. It might be better to keep the pre-existing graceful shutdown logic, and do a `while join_set.join_next().is_some()` to wait for shutdown to complete

**Comment by mwtian** [comment]
I was thinking that if canceling checkpoint creation is problematic, panic would cause inconsistency as well which I have not observed. The in-memory data seems per epoch and they do not seem to be concurrently accessed. I will add a fail point and if there is a strong preference, I will add the wait for processing to finish.

File: crates/sui-core/src/checkpoints/mod.rs [hunk 1]
```diff
@@ -1944,26 +1907,24 @@ impl CheckpointService {
...
             epoch_store.clone(),
             notify_aggregator.clone(),
-            exit_rcv,
             certified_checkpoint_output,
             state.clone(),
             metrics.clone(),
         );
 
-        spawn_monitored_task!(aggregator.run());
+        tasks.spawn(aggregator.run());
```
Comment (mystenmark) [inline crates/sui-core/src/checkpoints/mod.rs:1927]: we are also losing the spawn_monitored_task! here. should be easy to change the macro to take a joinset.

**Comment by mystenmark** [comment]
> I was thinking that if canceling checkpoint creation is problematic, panic would cause inconsistency as well which I have not observed. The in-memory data seems per epoch and they do not seem to be concurrently accessed. I will add a fail point and if there is a strong preference, I will add the wait for processing to finish.

The difference is that the panic wipes all in memory state so we would only be dealing with crash-recovery issues. Still, the fact that we only do this at the end of the epoch is convincing to me. Between that and the fact that simtest would be able to find problems with this, I think its okay as is.

**Comment by mwtian** [comment]
Updated to wait for checkpoint service tasks to shut down.

File: crates/sui-node/src/lib.rs [hunk 2]
```diff
@@ -1657,15 +1659,19 @@ impl SuiNode {
...
                 checkpoint_metrics,
                 sui_tx_validator_metrics,
             }) = self.validator_components.lock().await.take()
             {
                 info!("Reconfiguring the validator.");
-                // Stop the old checkpoint service.
-                drop(checkpoint_service_exit);
+                // Stop the old checkpoint service and wait for them to finish.
+                let _ = checkpoint_service_exit.send(());
+                while let Some(_result) = checkpoint_service_tasks.join_next().await {}
```
Comment (halfprice) [inline crates/sui-node/src/lib.rs:1670]: Do we want to add a timeout here in case something is wrong in the service to prevent it from stopping?

**Review by halfprice — APPROVED** [review]

File: crates/sui-core/src/checkpoints/mod.rs (hunk 1 above)
Comment (mwtian) [reply crates/sui-core/src/checkpoints/mod.rs]: Using monitored_future!() now.

File: crates/sui-node/src/lib.rs (hunk 2 above)
Comment (mwtian) [reply crates/sui-node/src/lib.rs]: Added a panic in tests if this times out.

**Comment by mwtian** [comment]
By wrapping `build.run()` with `epoch_store_clone.within_alive_epoch()`, IIUC it also does forces cancellation when the epoch ends. I'm not sure if the same wrapping needs to be applied to `aggregator.run()` as well. Will leave this for another PR if needed.

**Pull request merged**